### ISSUE TITLE
Metrics missing keys (?)

### ISSUE THREAD

**Issue by turtle0x1** [lxc/lxd#9322]
So according to the [spec](https://discuss.linuxcontainers.org/t/lxd-metric-exporter-for-instances/11735)

`lxd_cpu_seconds_total{cpu="<cpu>", mode="<mode>"}`

But if we analyse the output of [stgraber gist](https://gist.github.com/stgraber/ab7f204fb4bf53dbe134f6460bf41470) we can notice the `cpu` property is missing here https://gist.github.com/stgraber/ab7f204fb4bf53dbe134f6460bf41470#file-metrics-L259-L266

Ive also noticed the `type` property is missing from some entries in the following, but the spec doesn't say they are there so im not sure;
```
lxd_filesystem_avail_bytes
lxd_filesystem_free_bytes
lxd_filesystem_size_bytes
```
I dont know if this is a bug or not?

**Comment by monstermunchkin** [comment]
> So according to the [spec](https://discuss.linuxcontainers.org/t/lxd-metric-exporter-for-instances/11735)
> 
> `lxd_cpu_seconds_total{cpu="<cpu>", mode="<mode>"}`
> 
> But if we analyse the output of [stgraber gist](https://gist.github.com/stgraber/ab7f204fb4bf53dbe134f6460bf41470) we can notice the `cpu` property is missing here https://gist.github.com/stgraber/ab7f204fb4bf53dbe134f6460bf41470#file-metrics-L259-L266

That's not a bug. The CPU stats we get are for each CPU individually *and* a total of all. The latter obviously doesn't contain a `cpu` tag. I will update the spec to clarify this.

> 
> Ive also noticed the `type` property is missing from some entries in the following, but the spec doesn't say they are there so im not sure;
> 
> ```
> lxd_filesystem_avail_bytes
> lxd_filesystem_free_bytes
> lxd_filesystem_size_bytes
> ```
> 

Also not a bug. If we don't know the name of the filesystem, we omit it. Currently, LXD supports `btrfs`, `tmpfs`, `ext4`, `xfs`, `nfs`, and `zfs`. Which filesystem are you not seeing?

**Comment by monstermunchkin** [comment]
I've updated the [spec](https://discuss.linuxcontainers.org/t/lxd-metric-exporter-for-instances/11735).

**Comment by turtle0x1** [comment]
> That's not a bug. The CPU stats we get are for each CPU individually and a total of all. The latter obviously doesn't contain a cpu tag. I will update the spec to clarify this.

Looking at the gist that would imply its only true of VM's? as no `container` has entries have a similar entry but the VM does?

> Also not a bug. If we don't know the name of the filesystem, we omit it. Currently, LXD supports btrfs, tmpfs, ext4, xfs, nfs, and zfs. Which filesystem are you not seeing?

Looking a the gist https://gist.github.com/stgraber/ab7f204fb4bf53dbe134f6460bf41470#file-metrics-L381-L382 are the only 2 lines were the `type` key is missing in `lxd_filesystem_avail_bytes` .

Zabbix Prometheus parser is a bit weird, maybe im doing something wrong.

**Comment by monstermunchkin** [comment]
> > That's not a bug. The CPU stats we get are for each CPU individually and a total of all. The latter obviously doesn't contain a cpu tag. I will update the spec to clarify this.
> 
> Looking at the gist that would imply its only true of VM's? as no `container` has similar entries?

You're right, only VMs show the CPU seconds of all CPUs together (missing `cpu` tag). Containers will always list the CPU number.

> Looking a the gist https://gist.github.com/stgraber/ab7f204fb4bf53dbe134f6460bf41470#file-metrics-L381-L382 are the only 2 lines were the type key is missing in lxd_filesystem_avail_bytes .

Actually, LXD doesn't omit the fstype. If the name is not known, it'll only show the hexadecimal value.

**Comment by monstermunchkin** [comment]
I looked at the metrics (not the gist) when running containers and VMs, and cannot see that the `type` is missing. It's present everywhere.

**Comment by turtle0x1** [comment]
> > > That's not a bug. The CPU stats we get are for each CPU individually and a total of all. The latter obviously doesn't contain a cpu tag. I will update the spec to clarify this.
> > 
> > 
> > Looking at the gist that would imply its only true of VM's? as no `container` has similar entries?
> 
> You're right, only VMs show the CPU seconds of all CPUs together (missing `cpu` tag). Containers will always list the CPU number.

Not be that guy but does that make sense? Offering an aggregate for one type and not the other? Is it not better to have a consistent format which the user can aggregate if they want?
 
> > Looking a the gist https://gist.github.com/stgraber/ab7f204fb4bf53dbe134f6460bf41470#file-metrics-L381-L382 are the only 2 lines were the type key is missing in lxd_filesystem_avail_bytes .
> 
> ~Actually, LXD doesn't omit the fstype. If the name is not known, it'll only show the hexadecimal value.~ Sorry, I misread the problem. You are not seeing whether it's a VM or a container. That is definitely a bug.

> I looked at the metrics (not the gist) when running containers and VMs, and cannot see that the `type` is missing. It's present everywhere.

Maybe the gist is out of date, my `/1.0/metrics` endpoint doesn't work so I'm parsing the provided example :sweat_smile:

**Comment by monstermunchkin** [comment]
> Not be that guy but does that make sense? Offering an aggregate for one type and not the other? Is it not better to have a consistent format which the user can aggregate if they want?

We could to that. Since the aggregated value can be computed, I'd say we drop that value. @stgraber @tomponline what do you think?

> Maybe the gist is out of date, my /1.0/metrics endpoint doesn't work so I'm parsing the provided example

That could be. After the fixes are merged, your `/1.0/metrics` endpoint should work. In the meantime, what does your LXD log show? There should be an error regarding the metrics failure.

**Comment by turtle0x1** [comment]
I've updated #9323 to include the syslog entry (sorry forgot to paste the other night)
